modules/objdetect/test/test_qrcode_decode_encode.cpp

```diff
@@ -0,0 +1,222 @@
+            for (size_t i = 0; i < images_count; i++)
+            {
+                file_config << "{:" << "image_name" << qrcode_images_name[i];
+                std::string image_path = findDataFile(root + qrcode_images_name[i]);
+                std::vector<Point> corners;
+                Mat src = imread(image_path, IMREAD_GRAYSCALE), straight_barcode;
+                std::string decoded_info;
+                ASSERT_FALSE(src.empty()) << "Can't read image: " << image_path;
+
+                /**add the corner points*/
+                corners.push_back(Point(0,0));
+                corners.push_back(Point(src.rows-1,0));
+                corners.push_back(Point(src.rows-1,src.cols-1));
+                corners.push_back(Point(0,src.cols-1));
+
+                EXPECT_TRUE(decodeQRCode(src, corners, decoded_info, straight_barcode));
+
+                file_config << "x" << "[:";
+                for (size_t j = 0; j < corners.size(); j++) { file_config << corners[j].x; }
+                file_config << "]";
+                file_config << "y" << "[:";
+                for (size_t j = 0; j < corners.size(); j++) { file_config << corners[j].y; }
+                file_config << "]";
+                file_config << "info" << decoded_info;
+
+                file_config << "}";
+            }
+            file_config << "]";
+            file_config.release();
+        }
+#endif
+
+#ifdef UPDATE_ENCODE_TEST_DATA
+        TEST(Objdetect_QRCode, generate_test_data)
+        {
+            const std::string root = "qrcode/encode";
+            const std::string dataset_config = findDataFile(root + "dataset_config.json");
+            FileStorage file_config(dataset_config, FileStorage::WRITE);
+
+            file_config << "test_images" << "[";
+            size_t images_count = sizeof(qrcode_images_name) / sizeof(qrcode_images_name[0]);
+            for (size_t i = 0; i < images_count; i++)
+            {
+                file_config << "{:" << "image_name" << qrcode_images_name[i];
+                std::string image_path = findDataFile(root + qrcode_images_name[i]);
+                std::vector<Point> corners;
+                Mat src = imread(image_path, IMREAD_GRAYSCALE), straight_barcode;
+                std::string decoded_info;
+                ASSERT_FALSE(src.empty()) << "Can't read image: " << image_path;
+
+                /**add the corner points*/
+                corners.push_back(Point(0,0));
+                corners.push_back(Point(src.rows-1,0));
+                corners.push_back(Point(src.rows-1,src.cols-1));
+                corners.push_back(Point(0,src.cols-1));
+
+                int mode;
+                int version;
+                int ecc_level;
+                int mask_type;
+                int eci_num;
+
+                EXPECT_TRUE(decodeQRCode(src, corners, decoded_info, straight_barcode,mode,version,ecc_level,mask_type,eci_num));
```

Comment:
That's a good suggestion!

Reply:
But in this way , I can't get information including mode,version level ,ecc level etc. The only information I can get is decode_text . So that I can't have enough information to create a same qrcode ... Neither can I  test the image generated by my encoder and the orignial one pixel by pixel...

Reply:
please, follow the @allnes comment about `generate` method prototype (look at comments above)
follow this suggestion, `generate` method looks like
`bool generate(cv::String input, cv::OutputArray output, int version = 0, int correction_level = cv::CORRECT_LEVEL_L, int mode = cv::QR_MODE_AUTO, int structure_number = 1, Size output_size = (23, 23));`

So, all parameters except `input` can be calculated inside generating process